### PR TITLE
inthndlr: call 2F.1142 for redirector handles passed to 21.7142, call 2F.1142 in SftSeek2 if available  

### PR DESCRIPTION
**Pull request by ecm-pushbx**
Support for the new interface added in dosemu2's MFS https://github.com/dosemu2/dosemu2/pull/1376

### CODE REVIEW

**Comment by ecm-pushbx** [comment]
Builds fine but untested yet, due to my problems getting to run fdpp.

**Comment by ecm-pushbx** [comment]
Tested with `dosemu -I "lfn_support off"` now, works.

File: kernel/inthndlr.c [hunk 1]
```diff
@@ -1555,10 +1555,46 @@ VOID ASMCFUNC int21_service(iregs FAR * r)
 
       /* case 0x6d and above not implemented : see default; return AL=0 */
 
-    /* LFN API, support only 0xa6 for now */
+    /* LFN API, support only 0x42 and 0xA6 for now */
```
Comment (stsp) [inline kernel/inthndlr.c:1558]: Except that 0x42 is fat32 not lfn...
But whatever.

**Comment by stsp** [comment]
This misses the SEEK_CUR hook though.
And overall I am quite unhappy with it being
71 instead of 73, but I guess that's the life. :)

**Comment by ecm-pushbx** [comment]
I'm trying to write a wrapper around 1142h that goes into SftSeek. How do I allocate an array (two 32-bit integers) that I can take a far pointer off and pass to the redirector interface?

**Comment by stsp** [comment]
For what??
Can you just adjust SftSeek to call
1142 when needed?

**Comment by ecm-pushbx** [comment]
That's what I am trying. Here's my attempt so far, though I expect it doesn't compile yet: https://ulukai.org/ecm/20201225.3

The seek_struct must be allocated to DOS memory so I can initialise it, take a far pointer, pass it to 1142h, then read the result that 1142h wrote to the structure.

If 1142h is not supported for this handle, or the handle is local or a device, then SftSeek2 falls back to its original handling.

**Comment by stsp** [comment]
There are 2 ways of doing that.
Have a helper function in asm that
will take the far pointer and pass it
to 1142.
Exploit the fdpp-specific tricks, see
LBA_Get_Drive_Parameters()
as an example:
```
  bp = MK_FAR(lba_bios_parameters);
  regs.si = FP_OFF_OBJ(&regs, bp);
  regs.ds = FP_SEG_OBJ(&regs, bp);
```

IIRC you need to have bp and lba_bios_parameters
to have the similar life range:
```
  struct _bios_LBA_disk_parameterS lba_bios_parameters;
  struct _bios_LBA_disk_parameterS FAR *bp;
```

**Comment by stsp** [comment]
regs.si = FP_OFF_OBJ(&regs, bp);
is a way to tell fdpp code flow analyzer
that bp and regs are dependent, and
so accesses to "regs" may clobber "*bp".

**Comment by ecm-pushbx** [comment]
Finished it. Note to self: fdpp's "unsigned long" is a 64-bit type. I ended up using uint32_t in the structure definition to fix this.

File: kernel/dosfns.c [hunk 2]
```diff
@@ -271,6 +319,10 @@ COUNT SftSeek(int sft_idx, LONG new_pos, unsigned mode)
   {
     new_pos = 0;
   }
+  else if (try_long_compat_remote_seek(s, new_pos, mode, p_result))
+  {
+    return SUCCESS;
+  }
   else if (mode == SEEK_CUR)
   {
```
Comment (stsp) [inline kernel/dosfns.c:327]: Any reason you haven't made it
specific to SEEK_CUR?

File: kernel/dosfns.c [hunk 3]
```diff
@@ -255,6 +255,54 @@ long DosRWSft(int sft_idx, size_t n, __XFAR(void)bp, int mode)
...
+    *p_result = seek_struct.member[0];
+  }
+  else
+  {
+    *p_result = -1;
+  }
+  return 1;
+}
+
+COUNT SftSeek2(int sft_idx, LONG new_pos, unsigned mode, unsigned long * p_result)
```
Comment (stsp) [inline kernel/dosfns.c:305]: Please use freedos macros like
UWORD, UDWORD.

File: kernel/dosfns.c (hunk 2 above)
Comment (ecm-pushbx) [reply kernel/dosfns.c:327]: I think it is cleaner this way. And we don't depend on the redirector support for the 32-bit seek from EOF service. (I'm not certain whether that is correct in mfs now.) Also, for some corner cases it is better to always call into 1142h. Eg, what if the 32-bit seek ends up with sft_position = FFFFffffh?

File: kernel/proto.h [hunk 4]
```diff
@@ -83,6 +83,7 @@ BOOL check_break(void);
 UCOUNT GenericReadSft(__FAR(sft) sftp, UCOUNT n,__FAR(void) bp,
                       COUNT * err, BOOL force_binary);
 COUNT SftSeek(int sft_idx, LONG new_pos, unsigned mode);
+COUNT SftSeek2(int sft_idx, LONG new_pos, unsigned mode, unsigned long * p_result);
```
Comment (stsp) [inline kernel/proto.h:86]: Why?
Shouldn't it be static?

File: kernel/dosfns.c [hunk 5]
```diff
@@ -255,6 +255,54 @@ long DosRWSft(int sft_idx, size_t n, __XFAR(void)bp, int mode)
...
+  struct seek_struct_S FAR * ssp;
+  seek_struct.member[0] = new_pos;
+  if (mode == SEEK_SET)
+    seek_struct.member[1] = 0;
+  else if (0 == (new_pos & 0x80000000))
+    seek_struct.member[1] = 0;
+  else
+    seek_struct.member[1] = 0xFFFFffff;
+
+  iregs regs = {};
```
Comment (stsp) [inline kernel/dosfns.c:282]: Would be better to move that up to
the declarations.

File: kernel/dosfns.c (hunk 3 above)
Comment (ecm-pushbx) [reply kernel/dosfns.c:305]: Done.

File: kernel/dosfns.c [hunk 6]
```diff
@@ -255,6 +255,54 @@ long DosRWSft(int sft_idx, size_t n, __XFAR(void)bp, int mode)
...
+}
+
+static unsigned try_long_compat_remote_seek(sft FAR * s, LONG new_pos,
+	unsigned mode, unsigned long * p_result)
+{
+  if (0 == (s->sft_flags & SFT_FSHARED))
+    return 0;
+
+  struct seek_struct_S {
+    uint32_t member[2];
```
[the remark this answers is not in the document]
Comment (ecm-pushbx) [reply kernel/dosfns.c:270]: Done.

File: kernel/proto.h (hunk 4 above)
Comment (ecm-pushbx) [reply kernel/proto.h:86]: I didn't detect that DosSeek is in the same file. I'm about to amend the commit to make this static, yes.

File: kernel/dosfns.c (hunk 5 above)
Comment (ecm-pushbx) [reply kernel/dosfns.c:282]: Done.

File: kernel/proto.h (hunk 4 above)
Comment (ecm-pushbx) [reply kernel/proto.h:86]: Done.

**Comment by ecm-pushbx** [comment]
Something is odd here. My updates don't show up correctly.

**Comment by ecm-pushbx** [comment]
Got it, forgot the `-a` switch to `git commit --amend`

**Comment by stsp** [comment]
Thanks!
Please consider adding a test-case
for that functionality.

**Comment by andrewbird** [comment]
I was thinking about adding a test case for this on dosemu, but I'm not really clear about the calling convention and where it's expected to work (just MFS or FAT too?). I see the new function is int21/7142 but beyond that what input should I give it and what output should I expect? Does it work on small seeks too?

**Comment by stsp** [comment]
This is redirector-only thing.
You pass "file handle" in BX and
a pointer to 64bit offset in DS:DX.
CX is SEEK_SET/CUR/END.

**Comment by andrewbird** [comment]
Cool, thanks. And where do I see the resultant position returned?

**Comment by stsp** [comment]
At the same place where the original was.

**Comment by andrewbird** [comment]
Thanks

**Comment by andrewbird** [comment]
So is there anything I need to do to 'enable' the new 7142 command? In my test at https://github.com/andrewbird/dosemu2/commit/b97529ffac22e7c5a15dd9a7da8a47a5b3e17b92 I see that the new redirector call is used on normal djgpp lseeks

~~~
MFS: Entering dos_fs_redirect, FN=42, 'Long seek (extension)'                   
MFS: selecting drive                                                            
MFS: selected drive 23: /tmp/                                                   
MFS: long seek                                                                  
found /tmp/lfsfiles.tst on fd 37                                                
cl=00h seek=0000000Ah sftposition=00000000h                                     
fseek=00000000h fsize=180000000h                                                
result seek=0000000Ah                                                           
MFS: Finished dos_fs_redirect                                                   
~~~
but when I call the new 7142 command I get CF set and AX=7100 not implemented, and the call is never seen by the redirector.

Using FDPP with 
`python3 test/test_dos.py PPDOSGITTestCase.test_lfs_file_seek_tell_mfs_end`

**Comment by stsp** [comment]
I added a few blind/untested patches
to make it supported. As it was indeed
unsupported.

**Comment by andrewbird** [comment]
Thanks, I'll give it a try later today (just new year here!) Hope it will be a better one for all!

**Comment by andrewbird** [comment]
So that's a lot better, I'm now seeing the call to the redirector (that actually surprises me as I thought only SEEK_END would get passed to the redirector), but I'm seeing two calls. The final posn seems to be correct on the first, but something odd on the second.
~~~
MFS: Entering dos_fs_redirect, FN=42, 'Long seek (extension)'                   
MFS: selecting drive                                                            
MFS: selected drive 23: /tmp/                                                   
MFS: long seek                                                                  
found /tmp/lfsfiles.tst on fd 37                                                
cl=00h seek=0000000Ah sftposition=00000000h                                     
fseek=00000000h fsize=180000000h                                                
result seek=0000000Ah                                                           
MFS: Finished dos_fs_redirect                                                   
MFS: Entering dos_fs_redirect, FN=42, 'Long seek (extension)'                   
MFS: selecting drive                                                            
MFS: selected drive 23: /tmp/                                                   
MFS: long seek                                                                  
found /tmp/lfsfiles.tst on fd 37                                                
cl=00h seek=154B7B850E875D3h sftposition=0000000Ah                              
fseek=0000000Ah fsize=180000000h                                                
result seek=154B7B850E875D3h                                                    
MFS: Finished dos_fs_redirect                                                   
~~~

And the test that generated it
`python3 test/test_dos.py PPDOSGITTestCase.test_lfs_file_seek_tell_mfs_set`

**Comment by stsp** [comment]
You see 2 long seeks likely because
normal seek also goes there.

**Comment by andrewbird** [comment]
okay I'll investigate tomorrow why the return value is not what I expect, thanks!

**Comment by stsp** [comment]
Should now work.

**Comment by andrewbird** [comment]
Yes, those latest changes are good. Now works correctly with SEEKSET, SEEKCUR and SEEKEND correctly on FDPP, MSDOS 6.22 and DRDOS 7.01.

**Comment by stsp** [comment]
So your tests seems to be the only thing
that keeps our layered approach alive. :)
Without tests, every new addition misses
some of 10+ existing layers where it have
to be supported.
Maybe the initial ECMs patch, where the
whole thing was implemented in a single
place, is indeed easier to comprehend.

**Comment by ecm-pushbx** [comment]
> This is redirector-only thing.
> You pass "file handle" in BX and
> a pointer to 64bit offset in DS:DX.
> CX is SEEK_SET/CUR/END.

Sorry, I found this thread's updates only now.

The seek origin is actually (as I also commented on one of Andrew's commits) in CL, not CX.

And it is mostly redirector-only, the only other support currently known is Enhanced DR-DOS on a FAT+ FS partition. That originated the 7142h call.

**Comment by andrewbird** [comment]
> The seek origin is actually (as I also commented on one of Andrew's commits) in CL, not CX.

I have an update for my original comment in light of your clarification in an as yet unsubmitted PR

Thanks

**Pull request merged**